[Merged by Bors] - feat(data/polynomial/reverse): define `reverse f`, prove that `reverse` is a multiplicative monoid homomorphism

adomani opened the pull request:
---
the proof of `pol_ind_Rhom_prod_on_card` consists of two nested inductions.  Looking at the proofs, it is likely that the two inductions can be done separately, introducing an extra lemma and applying the lemma twice in the proof.  If there is substantial interest in trying this, I may give it a go and see if I can actually do it!
<!--
put comments you want to keep out of the PR commit here.
If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

adomani committed:
lemma showing that a polynomial whose support has cardinality at most one is a C_mul_X_pow

adomani committed:
define reverse, taking a polynomial to the polynomial with reversed coefficietns.  Prove that reverse is a multiplicative monoid homomorphism

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into poly-reverse
pulling in order to create a PR with reverse

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+# Reverse of a univariate polynomial
+
+The main definition is `reverse`.  Applying `reverse` to a polynomial `f(x)` produces the
+polynomial `x^f.nat_degree * f(1/x)`.
```
This matches notation in mathlib a bit better:
```suggestion
The main definition is `reverse`.  Applying `reverse` to a polynomial `f : polynomial R` produces the
polynomial with a reversed list of coefficients, equivalent to `X^f.nat_degree * f(1/X)`.
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+
+namespace polynomial
+
+open polynomial finsupp finset --erase_lead
```
If we're not using `erase_lead` here, we can delete it. If we really need to go back, we can always look in the Git history.
```suggestion
open polynomial finsupp finset
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+
+open polynomial finsupp finset --erase_lead
+
+
+variables {R : Type*} [semiring R] {f : polynomial R}
+
```
```suggestion
variables {R : Type*} [semiring R] {f : polynomial R}
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+
+namespace rev
+/-- rev_at is a function of two natural variables (N,i).  If i ≤ N, then rev_at N i returns N-i,
+otherwise it returns N.  Essentially, this function is only used for i ≤ N. -/
```
Did I correctly understand the purpose of `rev_at`?

```suggestion
/-- If `i ≤ N`, then `rev_at N i` returns `N - i`, otherwise it returns `i`.

Essentially, this function is only used for `i ≤ N`.
The advantage of `rev_at N i` over `N - i` is that `rev_at` is an involution.
-/
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+  unfold rev_at,
+  split_ifs,
+  refl,
+end
```
We don't need `split_ifs` if we already know which case it will be:
```suggestion
if_pos H
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+namespace rev
+/-- rev_at is a function of two natural variables (N,i).  If i ≤ N, then rev_at N i returns N-i,
+otherwise it returns N.  Essentially, this function is only used for i ≤ N. -/
+def rev_at (N : ℕ) : ℕ → ℕ := λ i : ℕ, ite (i ≤ N) (N-i) i
```
I use the rule of thumb to use less `λ` and more named parameters. (`simp` and `unfold` don't really care about this, but `rw` does).
```suggestion
def rev_at (N i : ℕ) : ℕ := ite (i ≤ N) (N-i) i
```

Vierkantor commented on src/data/polynomial/erase_lead.lean:
```diff
@@ -133,6 +133,23 @@ lemma erase_lead_nat_degree_lt (f0 : 2 ≤ f.support.card) :
 
+
+lemma C_mul_X_pow_of_card_support_le_one (h : f.support.card ≤ 1) :
+ f = C f.leading_coeff * X^f.nat_degree :=
```
```suggestion
  f = C f.leading_coeff * X ^ f.nat_degree :=
```

Vierkantor commented on src/data/polynomial/erase_lead.lean:
```diff
@@ -133,6 +133,23 @@ lemma erase_lead_nat_degree_lt (f0 : 2 ≤ f.support.card) :
+  by_cases f0 : f = 0,
+  { ext1,
+    rw [f0, leading_coeff_zero, C_0, zero_mul], },
+  { conv_lhs {rw ← erase_lead_add_C_mul_X_pow f},
```
```suggestion
  { conv_lhs { rw ← erase_lead_add_C_mul_X_pow f },
```

Vierkantor commented on src/data/polynomial/erase_lead.lean:
```diff
@@ -133,6 +133,23 @@ lemma erase_lead_nat_degree_lt (f0 : 2 ≤ f.support.card) :
+lemma C_mul_X_pow_of_card_support_le_one (h : f.support.card ≤ 1) :
+ f = C f.leading_coeff * X^f.nat_degree :=
+begin
+  by_cases f0 : f = 0,
```
I would say the second case is worth putting in a previous lemma ` C_mul_X_pow_of_card_support_eq_one (h : f.support.card = 1) : f = C f.leading_coeff * X ^ f.nat_degree`.

adomani replied:
Thank you for the suggestion: this is the result.

```
lemma C_mul_X_pow_of_card_support_eq_one (h : f.support.card = 1) : f = C f.leading_coeff * X ^ f.nat_degree :=
begin
  conv_lhs { rw ← erase_lead_add_C_mul_X_pow f },
  have fe0 : f.erase_lead = 0,
  { rw [← support_eq_empty, ← card_eq_zero],
    apply nat.eq_zero_of_le_zero (nat.lt_succ_iff.mp _),
    convert erase_lead_support_card_lt _,
    { exact h.symm, },
    { intro f0,
      rw [f0, support_zero, card_empty] at h,
      exact nat.no_confusion h, }, },
  rw [fe0, zero_add],
end

lemma C_mul_X_pow_of_card_support_le_one (h : f.support.card ≤ 1) :
 f = C f.leading_coeff * X ^ f.nat_degree :=
begin
  by_cases f0 : f.support.card = 0,
  { rw [card_eq_zero, support_eq_empty] at f0,
    ext1,
    rw [f0, leading_coeff_zero, C_0, zero_mul], },
  { apply C_mul_X_pow_of_card_support_eq_one,
    exact le_antisymm h (nat.one_le_of_lt (nat.pos_of_ne_zero f0)), },
end
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+is only used when `N` is at least as large as the degree of `f`.  Eventually, it will be used with
+`N` exactly equal to the degree of `f`.  -/
+def reflect : ℕ → polynomial R → polynomial R :=
+ λ N : ℕ, λ f : polynomial R, ⟨ image (rev_at N)  (f.support), λ i : ℕ, f.coeff (rev_at N i),
```
You should be able to simplify this definition by using `finsupp.emb_domain` (`rev_at N` is an injective function on `nat`, so composing it with a `finsupp` gives a `finsupp`.)

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+/-- The function `reflect` of a natural number `N` and a polynomial `f`, applies the function
+`rev_at` to the exponents of the terms appearing in the expansion of `f`.  In practice, `reflect`
+is only used when `N` is at least as large as the degree of `f`.  Eventually, it will be used with
+`N` exactly equal to the degree of `f`.  -/
```
```suggestion
/-- `reflect N f` is the polynomial such that `(reflect N f).coeff i = f.coeff (rev_at N i)`.
In other words, the terms with exponent `[0, ..., N]` now have exponent `[N, ..., 0]`.

 In practice, `reflect` is only used when `N` is at least as large as the degree of `f`.
 Eventually, it will be used with `N` exactly equal to the degree of `f`.  -/
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+  { refl, },
+end
+
+@[simp] lemma rev_at_small {N n : ℕ} (H : n ≤ N) : rev_at N n = N-n :=
```
Side note: should `n` be `i` here and at `rev_at_invol`, for consistency with the rest of the file?
```suggestion
@[simp] lemma rev_at_of_le {N n : ℕ} (H : n ≤ N) : rev_at N n = N - n :=
```

Vierkantor replied:
I see a few `@[simp]` lemmas that require proofs not occuring in the equalities: `rev_at_small`,  `reflect_mul`. I think those will not be picked up by the `simp` tactic because it cannot figure out what to fill in for the proofs. Have you tested that they work?

adomani replied:
I am afraid that I do not understand this question, and hence the answer is no, I guess.  I do not know how ```simp``` works, but my intention was to communicate to Lean that ```rev_at``` should make it think of splitting the two cases that appear in the definition.  What tests should I do?

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+@[simp] lemma reflect_zero {n : ℕ} : reflect n (0 : polynomial R) = 0 := rfl
+
+@[simp] lemma reflect_zero_iff {N : ℕ} (f : polynomial R) :
+ reflect N (f : polynomial R) = 0 ↔ f=0 :=
```
`f` can be implicit here because it appears on both sides of the `iff`.
```suggestion
@[simp] lemma reflect_eq_zero_iff {N : ℕ} {f : polynomial R} :
 reflect N (f : polynomial R) = 0 ↔ f = 0 :=
```

Vierkantor replied:
Could you also make sure that you indent the continuation lines for types with two spaces? For example on lines 73 and 84 there is only one space.

adomani replied:
I made ```f``` implicit.  I also added the spacing that you suggested.  I am not sure about the "indentation for continuation lines for types".  I took this to mean: when I break a line because it is longer than 100 chars, I will use two spaces for the continuation, rather than only one.  Let me know if you had something else in mind!

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+end
+
+@[simp] lemma reflect_add (f g : polynomial R) (n : ℕ) :
+ reflect n (f+g) = reflect n f + reflect n g :=
```
Should `n` be `N` here for consistency with the rest of the file?
```suggestion
@[simp] lemma reflect_add (f g : polynomial R) (n : ℕ) :
 reflect n (f + g) = reflect n f + reflect n g :=
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+begin
+  ext1,
+  unfold reflect,
+  rw [coeff_mk, coeff_C_mul, coeff_C_mul, coeff_mk],
```
For consistency with `coeff_C_mul`:
```suggestion
@[simp] lemma reflect_C_mul (f : polynomial R) (r : R) (N : ℕ) :
 reflect N (C r * f) = C r * (reflect N f) :=
begin
  ext1,
  unfold reflect,
  rw [coeff_mk, coeff_C_mul, coeff_C_mul, coeff_mk],
```

adomani replied:
I changed it here and in the rest of the file, thanks!

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+  ext1 m,
+  unfold reflect,
+  rw coeff_mk,
+  by_cases h : rev_at N n = m,
```
We should be able to skip a bit of rewriting if we first `rw [reflect_C_mul, coeff_C_mul, coeff_X_pow]`, then do `split_ifs`.

adomani replied:
I think that I followed your suggestion and obtained this proof

```
  ext,
  rw [reflect_C_mul, coeff_C_mul, coeff_C_mul, coeff_X_pow],
  unfold reflect,
  rw [coeff_mk],
  split_ifs,
  { rw [h, rev_at_invol, coeff_X_pow_self], },
  { rw [not_mem_support_iff_coeff_zero.mp],
    intro a,
    rw [← one_mul (X ^ n), ← C_1] at a,
    apply h,
    rw [← (mem_support_C_mul_X_pow a), rev_at_invol], },
```

I hope that this is what you had in mind!

Vierkantor replied:
That looks good, thanks! Maybe you can turn the `unfold reflect` into `rw reflect` as well, to save two more lines.

adomani replied:
Ah, I did not know that you could use ```rw``` instead of ```unfold```!  I used this trick a few more times to absorb ```unfolds``` into longer ```rw```s!

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+Even though this is not the actual definition, reverse f = f (1/X) * X ^ f.nat_degree. -/
+def reverse : polynomial R → polynomial R := λ f, reflect f.nat_degree f
+
+lemma pol_ind_Rhom_prod_on_card (cf cg : ℕ) {rp : ℕ → polynomial R → polynomial R}
```
Sorry, I don't understand the name and purpose of this lemma. It seems like `rp_add`, `rp_smul` and `rp_mon` uniquely characterise `reflect`, and we have already proven them by this point, so why not put `reflect` in the statement and inline the proofs?

adomani replied:
I think that you are right.  I arrived at this lemma after a long sequence of steps and once I could prove one of the versions, I simply went with what it was.  I had not realized that I had proved a "more general lemma" that was actually equivalent to what I wanted to prove!

adomani replied:
I will reformulate it, so that this is already the main statement.

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+
+@[simp] theorem reflect_mul
+ {f g : polynomial R} {F G : ℕ} (Ff : f.nat_degree ≤ F) (Gg : g.nat_degree ≤ G) :
+ reflect (F+G) (f*g) = reflect F f * reflect G g :=
```
```suggestion
 reflect (F + G) (f * g) = reflect F f * reflect G g :=
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+end
+
+theorem reverse_mul (f g : polynomial R) {fg : f.leading_coeff*g.leading_coeff ≠ 0} :
+ reverse (f*g) = reverse f * reverse g :=
```
Lean can't come up with `fg` by looking at the terms in the conclusion, so we have to make it explicit. Conversely, `f` and `g` appear in `fg` so we can make them implicit:
```suggestion
theorem reverse_mul {f g : polynomial R} (fg : f.leading_coeff * g.leading_coeff ≠ 0) :
 reverse (f * g) = reverse f * reverse g :=
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,214 @@
+  unfold reverse,
+  rw [nat_degree_mul' fg, reflect_mul (le_refl _) (le_refl _)],
+end
+
```
Could you add a version of `mul_reverse` that uses a typeclass argument to say there are no zero divisors? I think it should be enough to have the following assumptions:

```lean
@[simp] lemma reverse_mul_of_domain {R : Type*} [domain R] (f g : polynomial R) :
  reverse (f * g) = reverse f * reverse g := sorry
```

adomani replied:
I added the lemma and its proof.

```
@[simp] lemma reverse_mul_of_domain {R : Type*} [domain R] (f g : polynomial R) :
  reverse (f * g) = reverse f * reverse g :=
begin
  by_cases f0 : f=0,
  { rw [f0, zero_mul, reverse_zero, zero_mul], },
  { by_cases g0 : g=0,
    { rw [g0, mul_zero, reverse_zero, mul_zero], },
    { apply reverse_mul,
      apply mul_ne_zero;
      { rwa [← leading_coeff_eq_zero] at * }, }, },
end
```

Vierkantor reviewed (changes requested):
Thanks! :tada:

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Good catch!  The `N` was indeed a typo for `i`!

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into poly-reverse

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

I did not realize that it was a possibility not to use \lambda!

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/erase_lead.lean

This happens quite a few times: I will replace the remaining instances!

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/erase_lead.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
implementing the changes suggested by Vierkantor

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into poly-reverse

adomani committed:
Merge branch 'poly-reverse' of github.com:leanprover-community/mathlib into poly-reverse

adomani committed:
made a comment shorter than 100 chars

adomani committed:
shortened a line to fewer than 100 chars

bryangingechen commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,219 @@
+def reflect : ℕ → polynomial R → polynomial R :=
+  λ N : ℕ, λ f : polynomial R, ⟨ image (rev_at N)  (f.support), λ i : ℕ, f.coeff (rev_at N i),
+  begin
+
```
```suggestion
```

adomani committed:
commented the non-working version of rev_at, using the working one

adomani committed:
used Vierkantor's API for reflect

adomani committed:
Merge branch 'master' of github.com:leanprover-community/mathlib into poly-reverse

adomani committed:
also the induction part is fixed

adomani committed:
fixed documentation for rev_at_fun and added documentation for rev_at
removed to empty lines

adomani committed:
erased implied imports

adomani committed:
erased further implied imports

adomani committed:
removed paragraphs before imports
removed a "cyclic" import

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,206 @@
+  { exact nat.sub_sub_self h, },
+  { exact false.rec (N - i = i) (j (nat.sub_le N i)), },
+  { exact eq.refl i, },
+end
```
Since we have already proven this for `rev_at_fun`, this can just become:
```suggestion
rev_at_fun_invol
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,206 @@
+
+Essentially, this function is only used for `i ≤ N`.
+The advantage of `rev_at_fun N i` over `N - i` is that `rev_at_fun` is an involution.
+-/
```
Since the rest of the file uses `rev_at` rather than `rev_at_fun`, let's make that clear in the documentation.
```suggestion
/-- If `i ≤ N`, then `rev_at_fun N i` returns `N - i`, otherwise it returns `i`.

This is the map used by the embedding `rev_at`.
-/
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,206 @@
+  rw [← @rev_at_fun_invol N a, hab, rev_at_fun_invol],
+end
+
+/-- `rev_at N` is `rev_at_fun N` bundled as an embedding. -/
```
```suggestion
/-- If `i ≤ N`, then `rev_at N i` returns `N - i`, otherwise it returns `i`.

Essentially, this embedding is only used for `i ≤ N`.
The advantage of `rev_at N i` over `N - i` is that `rev_at` is an involution.
-/
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,206 @@
+def rev_at (N : ℕ) : function.embedding ℕ ℕ :=
+  { to_fun := λ i , (ite (i ≤ N) (N-i) i),
+    inj' := rev_at_fun_inj }
+
```
If we have multiple ways of writing an expression, it's a good idea to choose one as canonical. That way, we avoid having to duplicate our work for all the different ways to express the same thing. Writing out a `simp` lemma serves as kind of a Lean-readable documentation of which one we prefer.
```suggestion

/-- We prefer to use the bundled `rev_at` over unbundled `rev_at_fun`. -/
@[simp] lemma rev_at_fun_eq (N i : ℕ) : rev_at_fun N i = rev_at N i := rfl

```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,206 @@
+/-- `rev_at N` is `rev_at_fun N` bundled as an embedding. -/
+def rev_at (N : ℕ) : function.embedding ℕ ℕ :=
+  { to_fun := λ i , (ite (i ≤ N) (N-i) i),
+    inj' := rev_at_fun_inj }
```
The first line of a definition doesn't need to be indented.
```suggestion
{ to_fun := λ i , (ite (i ≤ N) (N-i) i),
  inj' := rev_at_fun_inj }
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,206 @@
+ In practice, `reflect` is only used when `N` is at least as large as the degree of `f`.
+ Eventually, it will be used with `N` exactly equal to the degree of `f`.  -/
+noncomputable def reflect (N : ℕ) (f : polynomial R) : polynomial R :=
+  finsupp.emb_domain (rev_at N) f
```
```suggestion
finsupp.emb_domain (rev_at N) f
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,206 @@
+begin
+  induction cf with cf hcf,
+  --first induction: base case
+  { induction cg with cg hcg,
```
I suspect it might be easier to `import data.polynomial.div` and use `refine f.rec_on_horner _ _ _` than do induction on the cardinality of the support. But I didn't try it out in much detail.

adomani replied:
Thank you for all the remaining suggestions: I liked them all!

I will try to simplify further the induction and look into ```rec_on_horner```.  I saw that it uses ```Sort```s, which means that it might take me some time to parse it.  Anyway, thanks a lot!

Vierkantor replied:
You're welcome, thanks for doing the PR!

`Sort`s are not too hard if you know what `Type u` means. `Sort 0` is `Prop` and `Sort (u + 1)` is `Type u`. Writing `Sort u` instead of `Type*` is a way to make definitions a bit more generic, allowing `Prop` and `Type` without re-defining anything. (In fact it's the other way around: `Prop` is defined as an abbreviation of `Sort 0` and `Type u` as an abbreviation of `Sort (u + 1)`.)

adomani replied:
Thanks for the explanation!  In the meantime, I just started playing around with the code that you suggested and it works **very** well!  Here is what the proof looks like, right now!

```
lemma reflect_mul_induction2 (cf cg : ℕ) :
 ∀ N O : ℕ, ∀ f g : polynomial R,
 f.support.card ≤ cf.succ → g.support.card ≤ cg.succ → f.nat_degree ≤ N → g.nat_degree ≤ O →
 (reflect (N + O) (f * g)) = (reflect N f) * (reflect O g) :=
begin
  intros N O f g cf cg,
  refine f.rec_on_horner _ _ _,
  { simp only [reflect_zero, zero_mul, eq_self_iff_true, forall_true_iff], },
  { intros df dg cp0 cg cg0 hg ec0,
    rw reflect_mul hg ec0, },
  { intros p p0 hh pX g0,
    rw reflect_mul pX g0, },
end
```

adomani replied:
I am in the process of cleaning up unnecessary hypotheses, while understanding the mechanics of this induction.

adomani replied:
Sorry, I messed up: the very short proof above *uses* ```reflect_mul```, since I was doing something silly.  I will try to see how it would work without tautologically proving it from itself!

adomani replied:
@Vierkantor , I have been trying to use ```rec_on_horner``` and it seems that it invariably wants me to prove the ```reflect_mul``` in the special case in which one of the two factors is a monomial.  This is of course true, provable (I had it in an earlier version), but not necessarily shorter than the current induction proof, I think.  (The special case involves understanding the relation between reflecting and multiplying by a monomial, that is, translating the support.)

Unless you have an idea on how to simplify this further, I am not sure that I prefer to use ```rec_on_horner```.  What do you think?

Vierkantor replied:
Hmm, then induction on the support sounds like the best idea. Let me see if I can do some lower-level refactoring to clean up the proof...

adomani replied:
That would be great: I will look carefully at the diff afterwards, to learn how to work with Lean better!

Vierkantor replied:
If we add the following helper lemma:

```lean
lemma rev_at_add {N O n o : ℕ} (hn : n ≤ N) (ho : o ≤ O) :
  rev_at (N + O) (n + o) = rev_at N n + rev_at O o :=
begin
  rcases nat.le.dest hn with ⟨n', rfl⟩,
  rcases nat.le.dest ho with ⟨o', rfl⟩,
  repeat { rw rev_at_le },
  { rw [add_assoc, add_left_comm n' o, ← add_assoc], simp },
  { exact le_add_right (le_refl _), },
  { exact le_add_right (le_refl _), },
  { exact add_le_add hn ho }
end
```

Then I can shorten it to:
```lean
begin
  induction cf with cf hcf,
  --first induction: base case
  { induction cg with cg hcg,
    -- second induction: base case
    { intros N O f g Cf Cg Nf Og,
      rw [C_mul_X_pow_of_card_support_le_one Cf, C_mul_X_pow_of_card_support_le_one Cg],
      simp only [mul_assoc, X_pow_mul, ← pow_add X, reflect_C_mul, reflect_monomial,
                 add_comm, rev_at_add Nf Og] },
    -- second induction: induction step
    { intros N O f g Cf Cg Nf Og,
      by_cases g0 : g = 0,
      { simp [g0] },

      rw [← erase_lead_add_C_mul_X_pow g, mul_add, reflect_add, reflect_add, mul_add, hcg, hcg];
        try { assumption },
      { exact le_add_left card_support_C_mul_X_pow_le_one },
      { exact (le_trans (nat_degree_C_mul_X_pow_le g.leading_coeff g.nat_degree) Og) },
      { exact nat.lt_succ_iff.mp (gt_of_ge_of_gt Cg (erase_lead_support_card_lt g0)) },
      { exact le_trans erase_lead_nat_degree_le Og } } },
  --first induction: induction step
  { intros N O f g Cf Cg Nf Og,
    by_cases f0 : f=0,
    { simp [f0] },

    rw [← erase_lead_add_C_mul_X_pow f, add_mul, reflect_add, reflect_add, add_mul, hcf, hcf];
       try { assumption },
    { exact le_add_left card_support_C_mul_X_pow_le_one },
    { exact (le_trans (nat_degree_C_mul_X_pow_le f.leading_coeff f.nat_degree) Nf) },
    { exact nat.lt_succ_iff.mp (gt_of_ge_of_gt Cf (erase_lead_support_card_lt f0)) },
    { exact (le_trans erase_lead_nat_degree_le Nf) } },
end
```

adomani replied:
Thanks a lot!  I like your suggestion and I have just pushed it!

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,206 @@
+    { rw [g0, mul_zero, reverse_zero, mul_zero], },
+    { apply reverse_mul,
+      apply mul_ne_zero;
+      { rwa [← leading_coeff_eq_zero] at * }, }, },
```
You don't necessarily need to make everything go to the right if you finish your first subgoal quickly. Basically in informal mathematics "if `f = 0` then we are done, if `g = 0` then we are done, otherwise ...".

```suggestion
  by_cases f0 : f=0,
  { rw [f0, zero_mul, reverse_zero, zero_mul], },
  by_cases g0 : g=0,
  { rw [g0, mul_zero, reverse_zero, mul_zero], },
  apply reverse_mul,
  apply mul_ne_zero;
    rwa [← leading_coeff_eq_zero] at *
```

Vierkantor reviewed (changes requested):
Thank you for the changes! It's starting to look quite nice. I would like to shorten `reflect_mul_induction` a bit more, and some other easier suggestions.

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Vierkantor's suggestions

adomani committed:
added induction temporary file

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,206 @@
+-/
+def rev_at_fun (N i : ℕ) : ℕ := ite (i ≤ N) (N-i) i
+
+@[simp] lemma rev_at_fun_invol {N i : ℕ} : rev_at_fun N (rev_at_fun N i) = i :=
```
The linting error we get is:
> `#print polynomial.rev.rev_at_fun_invol` /- simp can prove this:
> `by simp only [polynomial.rev.rev_at_invol, polynomial.rev.rev_at_fun_eq]`
> One of the lemmas above could be a duplicate.
> If that's not the case try reordering lemmas or adding @[priority].

That means we can drop the `@[simp]` tag:

```suggestion
lemma rev_at_fun_invol {N i : ℕ} : rev_at_fun N (rev_at_fun N i) = i :=
```

adomani committed:
half works

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
check the merge with mathlibe

adomani committed:
Merge branch 'poly-reverse' of github.com:leanprover-community/mathlib into poly-reverse

adomani committed:
added the main file, that I removed by mistake

adomani committed:
added a removed file

adomani committed:
Changes to be committed:
	deleted:    src/data/polynomial/induction_horner.lean

adomani committed:
implemented improvement on induction using rev_at_add by Vierkantor

adomani committed:
forgot to pull the PR

adomani committed:
Merge branch 'poly-reverse' of github.com:leanprover-community/mathlib into poly-reverse

adomani committed:
clean up and formatting

adomani committed:
with to_reverse

adomani committed:
without to_reverse

adomani committed:
proof that trailing_coeff is a multiplicative hom

adomani committed:
local lint modifications

adomani committed:
clean up of to_reverse

adomani committed:
moved to_reverse to a branch poly-reverse2

adomani committed:
Merge branch 'master' into poly-reverse

adomani committed:
Update reverse.lean

replaced ```C_mul_X_pow_of_card_support_le_one``` with the already existing ```C_mul_X_pow_eq_self```

adomani committed:
version with two lemmas on card.support \leq 1

adomani committed:
Merge branch 'poly-reverse' of github.com:leanprover-community/mathlib into poly-reverse

jcommelin reviewed (approved):
LGTM, but I will wait for @Vierkantor

kim-em commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,203 @@
+
+variables {R : Type*} [semiring R] {f : polynomial R}
+
+namespace rev
```
Why is this in a namespace `rev`? I would omit this.

adomani replied:
I removed the namespace in my local file!  It will get uploaded once I push!

kim-em commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,203 @@
+
+@[simp] lemma reflect_add (f g : polynomial R) (N : ℕ) :
+  reflect N (f + g) = reflect N f + reflect N g :=
+by { ext1, rw [coeff_add, coeff_reflect, coeff_reflect, coeff_reflect, coeff_add], }
```
Do proofs like this not work by `by { ext, simp }`? If `simp` handles it (which I'd hope it should, if you have enough simp lemmas), it just seems clutter to spell out the steps.

jcommelin replied:
Wouldn't we want to `squeeze_simp` the result anyway? I've seen quite a bunch of slowish proofs recently that could be sped up using `squeeze_simp`.
This brings me back to my point of wishing for two layers of proof presentations: the once that show how Lean can *find* a proof itself, and those where Lean only really has to do verification (even if it uses tactics).

adomani replied:
You are right that ```{ ext, simp }``` completes the proof, but I strive as much as possible to avoid using ```simp```.  If you believe that this is a mistake, I can leave ```simp```s in proofs, but I was explicitly replacing ```simp```s by step-by-step proofs.

Vierkantor replied:
A quick, unscientific experiment using `set_option profiler true` shows the following times on my machine:

 * `by { ext1, rw }`: 9 ms tactic execution
 * `by { ext, simp only [coeff_add, coeff_reflect] }`: 22 ms tactic execution
 * `by { ext, simp }`: 231 ms tactic execution

231 milliseconds is already large enough to notice, so I would vote for keeping them as `rw` or, as a compromise, using `simp only`.

adomani replied:
I am very happy to have learnt about ```set_option profiler true``` and that it supports my coding idiosyncrasies...

adomani replied:
Still, I am happy to comply with alternative conventions, should this be preferred!

jcommelin replied:
If a `rw` mentions some lemmas multiple times, I usually opt for the `simp only` for brevity.

adomani committed:
removed namespace rev, following Scott Morrison's suggestion

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,202 @@
+
+@[simp] lemma reflect_add (f g : polynomial R) (N : ℕ) :
+  reflect N (f + g) = reflect N f + reflect N g :=
+by { ext1, rw [coeff_add, coeff_reflect, coeff_reflect, coeff_reflect, coeff_add], }
```
```suggestion
by { ext, simp only [coeff_add, coeff_reflect], }
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,202 @@
+
+@[simp] lemma reflect_C_mul (f : polynomial R) (r : R) (N : ℕ) :
+  reflect N (C r * f) = C r * (reflect N f) :=
+by { ext1, rw [coeff_reflect, coeff_C_mul, coeff_C_mul, coeff_reflect], }
```
```suggestion
by { ext, simp only [coeff_reflect, coeff_C_mul], }
```

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,202 @@
+  reflect N (C c * X ^ n) = C c * X ^ (rev_at N n) :=
+begin
+  ext,
+  rw [reflect_C_mul, coeff_C_mul, coeff_C_mul, coeff_X_pow, coeff_reflect],
```
```suggestion
  simp only [reflect_C_mul, coeff_C_mul, coeff_X_pow, coeff_reflect],
```

adomani replied:
@Vierkantor , @jcommelin 
I reverted this ```simp only``` to a ```rw```: this is what caused the CI failure.  It was the only "non-final" ```simp``` and the rest of the proof depended on the exact output of the ```rw```: ```simp only``` went too far!

Vierkantor replied:
Ah, oops! I must have overlooked the error when I tried it out.

adomani replied:
No problem!  As always, thank you for your help!

Vierkantor commented on src/data/polynomial/reverse.lean:
```diff
@@ -0,0 +1,202 @@
+  reverse (f * g) = reverse f * reverse g :=
+begin
+  by_cases f0 : f=0,
+  { rw [f0, zero_mul, reverse_zero, zero_mul], },
```
```suggestion
  { simp only [f0, zero_mul, reverse_zero], },
```

Vierkantor reviewed (approved):
Since the consensus seems to favour `simp only` over `rw`, these would be my suggestions and then I'm happy. Thanks for all the work!

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
Update src/data/polynomial/reverse.lean

Co-authored-by: Anne Baanen <Vierkantor@users.noreply.github.com>

adomani committed:
reverted a non-final ```simp only``` to a ```rw```
